feat: add environment variables validation

piotrski opened the pull request:
# Add runtime environment variables validation

- [x] I reviewed linter warnings + errors, resolved formatting, types and other issues related to my work
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

In #140 @KATT suggested adding runtime environment variables validation using zod. Since `env.js` is imported inside `next.config.js` it can't ba a typescript file. There are two templates for `env.js`, one for just prisma `env-prisma.js` and another one for prisma + next-auth `env-prisma-auth.js`

---

## Screenshots
<img width="639" alt="Screenshot 2022-07-07 at 20 58 14" src="https://user-images.githubusercontent.com/244174/177853783-e7927ef2-3e28-4454-8746-765c1b566763.png">

💯

piotrski commented:
> On my phone, but think you should add something to the tsconfig too in order to get the file validated with ts ☺

I've added `**/*.js` to tsconfig.

piotrski commented:
> Hi there! Please remember to update the PR title so we get it on the changelog.
> 
> feat: add environment variables validation
> 
> Cheers!

Should I change it to something else?

Fedeorlandau commented:
> > Hi there! Please remember to update the PR title so we get it on the changelog.
> > feat: add environment variables validation
> > Cheers!
> 
> Should I change it to something else?

It's perfect now! Thanks

Quick question, what happens when just "next-auth + trpc" or just "next-auth"  is enabled?

TheoBr reviewed (approved):
I do really like this

juliusmarminge commented:
The only problem with this as it stands, is that it would make the initial app fail to build because of the example .env not following this validation-schema. 

I just added a CI step that builds the default t3-app and this would fail if we do not include a better `.env` by default that can be parsed with this validation.

juliusmarminge commented:
> There are two templates for env.js, one for just prisma env-prisma.js and another one for prisma + next-auth env-prisma-auth.js

Also include one for just NextAuth.js `env-auth.js` to cover all the cases.

piotrski commented:
I've moved schema to `env-schema.js` to keep the code DRY and added env validation to the base template. Now addons are just need to override `env-schema.js`. I've tested **base config**, **prisma**, **prisma + auth**, **auth**, **auth + tailwind**

juliusmarminge commented on src/installers/next-auth.ts:
```diff
@@ -38,8 +38,12 @@ export const nextAuthInstaller: Installer = async ({
   );
 
+  const envSchemaSrc = path.join(nextAuthAssetDir, "env-auth.js");
+  const envSchemaDest = path.join(projectDir, "src/server/env-schema.js");
```
you should probably do some more checking to make sure the correct schema is selected. 

as it is now, are you sure that the Prisma-installer will overwrite this and select the `env-prisma-auth.js` _after_ this? if it runs the other way around this will override in the other direction if that makes sense?

_maybe_ having a separate installer for this would make handling this logic easier?

piotrski replied:
I could move it to a separate installer for now. We only have 3 variants, this should be manageable

juliusmarminge replied:
yes, until there are other stuff being made with ast/codeshift stuff there is no reason to do that for this one

juliusmarminge commented on src/installers/prisma.ts:
```diff
@@ -41,6 +41,12 @@ export const prismaInstaller: Installer = async ({
   const sampleApiRouteSrc = path.join(prismaAssetDir, "sample-api.ts");
   const sampleApiRouteDest = path.join(projectDir, "src/pages/api/examples.ts");
 
+  const envSchemaSrc = path.join(
```
same as above

juliusmarminge commented on template/addons/next-auth/api-handler-prisma.ts:
```diff
@@ -5,14 +5,15 @@ import CredentialsProvider from "next-auth/providers/credentials";
 // Prisma adapter for NextAuth, optional and can be removed
 import { PrismaAdapter } from "@next-auth/prisma-adapter";
 import { prisma } from "../../../server/db/client";
+import { env } from "../../../server/env";
```
in your installers you copy this file to `server/env-schema` or am I missing something?

piotrski replied:
I decided to split it into two separate files to keep it dry
- `env-schema.js` - contains the schama overwritten by the installers
- `env.js` - parses schema, formats errors and exports validated env variables

juliusmarminge commented on template/addons/prisma/env-prisma-auth.js:
```diff
@@ -0,0 +1,12 @@
+  GITHUB_SECRET: z.string(),
+});
+
+module.exports.envSchema = envSchema;
```
these can not be ESM, right?
```js
export const envSchema = ...
```

piotrski replied:
Yes, but I think the `.mjs` might be supported

juliusmarminge reviewed (changes requested):
I like this. Left some comments. As for the concern in my comment previously, would this prevent the app from building if the .env schema is not valid or is it just runtime?

juliusmarminge commented on .github/workflows/PR-CI.yml:
```diff
@@ -95,4 +95,5 @@ jobs:
       # has to be scaffolded outside the CLI project so that no lint/tsconfig are leaking
       # through. this way it ensures that it is the app's configs that are being used
       - run: pnpm start -y ../ci-test-app
+      - run: mv ../ci-test-app/.env-example ../ci-test-app/.env
```
oh I like this approach. this way we still leave it up to the user to make sure they set their env-vars whilst allowing the CI to test-build the app 👍🏼

juliusmarminge commented:
Looks good! Update the branch and I'll merge

piotrski commented:
> Looks good! Update the branch and I'll merge

Updated 👍

juliusmarminge commented:
Thank you 🙏!
